src/commissioner/States.h

```diff
@@ -0,0 +1,840 @@
+{
+    AwaitingCommissionableDiscovery(const CommissionableNodeDiscovery<TContext> & previous) :
+        CommissionableNodeDiscovery<TContext>(previous)
+    {
+        this->mName = "AwaitingCommissionableDiscovery";
+    }
+    void Enter() { this->mDiscoverer.get()->SetDelegate(this); }
+
+protected:
+    void OnDiscovery() override { this->mCtx.Dispatch(TContext::Event::template Create<Events::Success>()); }
+    void OnDiscovererShutdown() override {}
+};
+
+template <typename TContext>
+struct PasscodeAuthenticatedSessionEstablishment : CommissionableNodeDiscovery<TContext>, SessionEstablishmentDelegate
+{
+    PasscodeAuthenticatedSessionEstablishment(const CommissionableNodeDiscovery<TContext> & previous) :
+        CommissionableNodeDiscovery<TContext>(previous)
+    {
+        this->mName = "PasscodeAuthenticatedSessionEstablishment";
+    }
+
+    void Enter()
+    {
+        this->mDiscoverer.get()->SetDelegate(this);
+        while ((mStatus = this->mDiscoverer.get()->GetNextCandidate(this->mCommissionee)) == CHIP_NO_ERROR)
+        {
+            VerifyOrReturn((mStatus = TryPase()) != CHIP_NO_ERROR);
+        }
+        if (mStatus == CHIP_ERROR_NOT_FOUND)
+        {
+            // CHIP_ERROR_NOT_FOUND from the discoverer means we should keep looking.
+            // This is not ncessarily a failure.
+            this->mCtx.Dispatch(TContext::Event::template Create<Events::Await>());
+        }
+        else
+        {
+            this->mDiscoverer.get()->Shutdown();
+        }
+    }
+
+    void ShutdownDiscoverer() { this->mDiscoverer.get()->Shutdown(); }
+
+protected:
+    CHIP_ERROR TryPase()
+    {
+        CHIP_ERROR err                        = CHIP_NO_ERROR;
+        uint16_t sessionId                    = 0;
+        Messaging::ExchangeContext * exchange = nullptr;
+        Optional<SessionHandle> session;
+
+        mPairing = Platform::MakeShared<PASESession>();
+        VerifyOrExit(mPairing.get() != nullptr, err = CHIP_ERROR_NO_MEMORY);
+        session = this->mCommissionee.mSystemState->SessionMgr()->CreateUnauthenticatedSession(
+            this->mCommissionee.mCommissionableNodeAddress.Value(), this->mCommissionee.mMrpConfig.ValueOr(gDefaultMRPConfig));
+        VerifyOrExit(session.HasValue(), err = CHIP_ERROR_NO_MEMORY);
+        exchange = this->mCommissionee.mSystemState->ExchangeMgr()->NewContext(session.Value(), this->mPairing.get());
+        VerifyOrExit(exchange != nullptr, err = CHIP_ERROR_INTERNAL);
+        SuccessOrExit(err = this->mCommissionee.mSystemState->GetSessionIDAllocator()->Allocate(sessionId));
+        SuccessOrExit(err = mPairing.get()->Pair(this->mCommissionee.mCommissionableNodeAddress.Value(),
+                                                 this->mPayload.get()->setUpPINCode, sessionId, this->mCommissionee.mMrpConfig,
+                                                 exchange, this));
+    exit:
+        return err;
```

Comment:
Would you need to destroy `exchange` or reset `mPairing` in `exit:`?

Reply:
Ah, very interesting.  There is a potential resource leak, both here and in CHIPDeviceController.  And it isn't quite the leak I thought:

From CHIPDeviceController (the problem is the same in both pieces of code):

```
    // TODO: In some cases like PASE over IP, CRA and CRI values from commissionable node service should be used
    session = mSystemState->SessionMgr()->CreateUnauthenticatedSession(params.GetPeerAddress(), device->GetMRPConfig());
    VerifyOrExit(session.HasValue(), err = CHIP_ERROR_NO_MEMORY);

    exchangeCtxt = mSystemState->ExchangeMgr()->NewContext(session.Value(), &device->GetPairing());
    VerifyOrExit(exchangeCtxt != nullptr, err = CHIP_ERROR_INTERNAL);

    err = mIDAllocator.Allocate(keyID);
    SuccessOrExit(err);

    err = device->GetPairing().Pair(params.GetPeerAddress(), params.GetSetupPINCode(), keyID,
                                    Optional<ReliableMessageProtocolConfig>::Value(mMRPConfig), exchangeCtxt, this);
```

Pair() will free exchangeCtxt on error, but it can only do this obviously if it is actually called.  So in the case that Allocate fails, this code is actually leaking an exchange context.

It is likely that a similar bug exists in the OperationalDeviceProxy.

I will fix the issue here and file other issues as appropriate.  Thank you for the careful review.

Reply:
Filed here: https://github.com/project-chip/connectedhomeip/issues/13422